### PR TITLE
Remove HIR inlining

### LINKED ISSUE #49690: Remove HIR inlining (now that we don't need it for constant evaluation anymore)
Now that we have MIRI-based constant evaluation, we should be able to remove cross-crate HIR inlining and thus a source of mutability in the HIR map. This is needed for query parallelization.

### PR DESCRIPTION
**Pull request by wesleywiser**
Fixes #49690 

r? @michaelwoerister

### CODE REVIEW

**Comment by michaelwoerister** [comment]
Nice, thanks @wesleywiser! I'll take a look as soon as I find the time.

**Comment by michaelwoerister** [comment]
Looks great, @wesleywiser! Two things I'm wondering:
- Does this code also need to handle statics?
- Could the [schema::Entry::ast field](https://github.com/rust-lang/rust/blob/d6a2dd9912e762fb0029c4463002d674686c1159/src/librustc_metadata/schema.rs#L269) be removed entirely?

File: src/librustc_metadata/schema.rs [hunk 1]
```diff
@@ -375,14 +376,21 @@ impl<'a, 'gcx> HashStable<StableHashingContext<'a>> for EntryKind<'gcx> {
...
                 associated_container.hash_stable(hcx, hasher);
                 qualif.hash_stable(hcx, hasher);
             }
         }
     }
 }
 
+#[derive(RustcEncodable, RustcDecodable)]
+pub struct ConstData {
+    pub rendered: String,
```
Comment (oli-obk) [inline src/librustc_metadata/schema.rs]: Is this expected to gain more fields? If no, how about making t a newtype called `RenderedConst`?

Also 🚨documentation police🚨 please add some docs

File: src/librustc/ty/maps/mod.rs [hunk 2]
```diff
@@ -254,9 +253,9 @@ define_maps! { <'tcx>
     [] fn item_attrs: ItemAttrs(DefId) -> Lrc<[ast::Attribute]>,
     [] fn trans_fn_attrs: trans_fn_attrs(DefId) -> TransFnAttrs,
     [] fn fn_arg_names: FnArgNames(DefId) -> Vec<ast::Name>,
+    [] fn rendered_const: RenderedConst(DefId) -> String,
```
Comment (oli-obk) [inline src/librustc/ty/maps/mod.rs]: I know this is essentially self explanatory. Still, a short doc comment would be awesome

**Review by oli-obk — COMMENTED** [review]
👍

**Comment by wesleywiser** [comment]
@michaelwoerister 

> Does this code also need to handle statics?

I don't believe so. I was unable to find a failing test case for statics and all of the existing rustdoc tests pass after handling `Const` and `AssociatedConst`.

> Could the schema::Entry::ast field be removed entirely?

I think I ran into some issues when I tried that but I don't remember clearly. I will investigate that tonight.

File: src/librustc_metadata/schema.rs (hunk 1 above)
Comment (wesleywiser) [reply src/librustc_metadata/schema.rs]: I'm not aware of any additional data we need at this time. I'd be happy to change it and I'll add some docs too 👍

**Comment by wesleywiser** [comment]
`schema::Entry::ast` is still used by the [`typeck_tables_of` query](https://github.com/rust-lang/rust/blob/49317cd511fbb60178bd5122e484609568938468/src/librustc_metadata/cstore_impl.rs#L145) which calls [`item_body_tables`](https://github.com/rust-lang/rust/blob/7360d6dd678d186d9c9b46311b75ba6840e61aa2/src/librustc_metadata/decoder.rs#L812) and the [`const_is_rvalue_promotable_to_static` query](https://github.com/rust-lang/rust/blob/49317cd511fbb60178bd5122e484609568938468/src/librustc_metadata/cstore_impl.rs#L167) which calls a [function of the same name](https://github.com/rust-lang/rust/blob/7360d6dd678d186d9c9b46311b75ba6840e61aa2/src/librustc_metadata/decoder.rs#L842).

**Comment by wesleywiser** [comment]
Added docs & newtype wrapper.

**Comment by michaelwoerister** [comment]
> I don't believe so. I was unable to find a failing test case for statics and all of the existing rustdoc tests pass after handling Const and AssociatedConst.

:+1: 

> schema::Entry::ast is still used by the typeck_tables_of query which calls item_body_tables and the const_is_rvalue_promotable_to_static query which calls a function of the same name.

You might be able to remove the metadata-based version of the `typeck_tables_of` query, since it shouldn't be called by for external items anymore. About `const_is_rvalue_promotable_to_static` I'm not sure. @eddyb or @oli-obk might know more.

**Comment by oli-obk** [comment]
You'd need to move the `rvalue_promotable_tovstatic` field up from the `Ast` to  the `Entry`

File: src/librustc/middle/cstore.rs [hunk 3]
```diff
@@ -198,16 +198,6 @@ pub trait MetadataLoader {
...
-    pub body: &'tcx hir::Body,
-
-    // It would require a lot of infrastructure to enable stable-hashing Bodies
-    // from other crates, so we hash on export and just store the fingerprint
-    // with them.
-    pub fingerprint: ich::Fingerprint,
-}
-
 #[derive(Clone)]
 pub struct ExternBodyNestedBodies {
```
Comment (eddyb) [inline src/librustc/middle/cstore.rs]: Can this be removed too?

File: src/librustc_metadata/schema.rs [hunk 4]
```diff
@@ -289,7 +289,7 @@ impl_stable_hash_for!(struct Entry<'tcx> {
 
 #[derive(Copy, Clone, RustcEncodable, RustcDecodable)]
 pub enum EntryKind<'tcx> {
-    Const(u8),
+    Const(u8, Lazy<RenderedConst>),
```
Comment (eddyb) [inline src/librustc_metadata/schema.rs]: You could replace the `u8` here and in `AssociatedConst` with a `struct ConstQualif { mir: u8, ast_promotable: bool }`.

File: src/librustc/middle/cstore.rs (hunk 3 above)
Comment (wesleywiser) [reply src/librustc/middle/cstore.rs]: Yes, good catch!

File: src/librustc_metadata/schema.rs (hunk 4 above)
Comment (wesleywiser) [reply src/librustc_metadata/schema.rs]: What does `ast_promotable` mean?

Comment (eddyb) [reply src/librustc_metadata/schema.rs]: The `rvalue_promotable_to_static: bool` field that's currently in `Ast`.

**Comment by wesleywiser** [comment]
Implemented feedback & rebased.

File: src/librustc_metadata/encoder.rs [hunk 5]
```diff
@@ -908,20 +924,20 @@ impl<'a, 'b: 'a, 'tcx: 'b> IsolatedEncoder<'a, 'b, 'tcx> {
...
-        } else {
-            (None, false)
-        };
+        let mir =
+            if let hir::ImplItemKind::Const(..) = ast_item.node {
+                true
+            } else if let hir::ImplItemKind::Method(ref sig, _) = ast_item.node {
+                let generics = self.tcx.generics_of(def_id);
+                let types = generics.parent_types as usize + generics.types.len();
+                let needs_inline = types > 0 || tcx.trans_fn_attrs(def_id).requests_inline() &&
```
Comment (varkor) [inline src/librustc_metadata/encoder.rs]: Was this change intentional? This changes the behaviour as `&&` binds more tightly — was the previous behaviour incorrect?

Comment (michaelwoerister) [reply src/librustc_metadata/encoder.rs]: Also, this would be nicer with a `match` instead of chained `if let`s. @wesleywiser, would you mind making a PR with a fix?

Comment (wesleywiser) [reply src/librustc_metadata/encoder.rs]: Yeah, looks like I messed this up when I pulled the latest changes into my branch. I've opened a quick PR that fixes this (#50114). I'll send a follow up tonight that uses a match for this instead.

File: src/librustc_metadata/schema.rs [hunk 6]
```diff
@@ -375,14 +373,37 @@ impl<'a, 'gcx> HashStable<StableHashingContext<'a>> for EntryKind<'gcx> {
             EntryKind::AssociatedType(associated_container) => {
                 associated_container.hash_stable(hcx, hasher);
             }
-            EntryKind::AssociatedConst(associated_container, qualif) => {
+            EntryKind::AssociatedConst(associated_container, qualif, _) => {
```
Comment (eddyb) [inline src/librustc_metadata/schema.rs]: This ignores the last field, but `EntryKind::Const` above doesn't. cc @michaelwoerister

File: src/librustc_metadata/encoder.rs [hunk 7]
```diff
@@ -869,6 +871,13 @@ impl<'a, 'b: 'a, 'tcx: 'b> IsolatedEncoder<'a, 'b, 'tcx> {
         !self.tcx.sess.opts.output_types.should_trans()
     }
 
+    fn const_qualif(&self, mir: u8, body_id: hir::BodyId) -> ConstQualif {
```
Comment (eddyb) [inline src/librustc_metadata/encoder.rs]: This should ideally call `mir_const_qualif` itself instead of taking the value as an argument.
